pkg/apk/impl/rwfs/rwfs.go

```diff
@@ -0,0 +1,44 @@
+// Copyright 2022 Chainguard, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rwfs
+
+import (
+	"io"
+	"io/fs"
+)
+
+// FS is an interface for a filesystem. It includes fs.FS, as well as methods
+// for writing to files. It extends fs.FS since the fs.FS interface only
+// provides read-only.
+type FS interface {
+	fs.FS
+	fs.ReadDirFS
+	fs.StatFS
+	fs.ReadFileFS
+	OpenFile(name string, flag int, perm fs.FileMode) (File, error)
+	Mkdir(string, fs.FileMode) error
+	MkdirAll(string, fs.FileMode) error
+	WriteFile(string, []byte, fs.FileMode) error
+	Mknod(path string, mode uint32, dev int) error
+	Symlink(oldname, newname string) error
```

Comment:
We definitely need the ability to track permissions and ownership somehow, to ensure apko does not need to run as root.  This was the main reason for `pkg/vfs`.

Reply:
The approach I took was to track all target permissions and ownership at each stage, then add it to the tar file later as overrides. That definitely works. But, at some point I did think about doing away with that and replacing it with the vfs/rwfs layer and have that be responsible for tracking. Even if it is just wrapping memfs or os.DirFS or anything else, let it handle the above.

I almost did it, but had some concerns. First, I wanted everything else to be in the right place and looking good before revamping that part. Second, I was concerned about resource usage. The FS likely would need to track it all in memory, which can get big. It shouldn't get too big, but it can get big. Or it can add a "magic file" somewhere in its filesystem, but I don't like that much. Third, I was worried that if it stopped in the middle, we wouldn't be able to pick it up again.

But then I realized that I already am putting some things in memory, and premature optimization is not something to worry about - when it gets really resource hungry, we can look at alternatives, not before - and we pretty much clean it all up if we stop in the middle anyways. So why not?

To take an approach: are you suggesting we switch from my current implemented approach of, "reconstruct what we need" (based on permissions in `installed`, home dir overrides, etc.) to "have the FS track it all"?

Reply:
Either approach works.  If you have full support for the apkdb, that seems fine to me!

Reply:
If you are ok with it, I am going to punt on that change for a bit. I do like the elegance of doing it all in FS, but what we have now works. If everything else is ok, when we are done - either as part of this PR or as a follow-on - I am happy to switch it.